### ISSUE TITLE
compiler crash in iree-util-ipo pass

### ISSUE THREAD

**Issue by astrotuna201** [iree-org/iree#11062]
### What happened?

I tried to run a tool that uses jax with iree backend, leading to a compiler crash with reproducer as attached.

The crash is an assert in Block.cpp:

"cannot insert arguments to blocks with predecessors"

### Steps to reproduce your issue

See debug / reproducer output attached:

```
(lldb) target create "___/iree-install/bin/iree-compile"
Current executable set to '___iree-install/bin/iree-compile' (x86_64).
(lldb) settings set -- target.run-args  "core-input.mlir" "--iree-input-type=mhlo" "--iree-vm-bytecode-module-output-format=flatbuffer-binary" "--iree-hal-target-backends=llvm-cpu" "--iree-llvm-embedded-linker-path=/Users/wiggles/Library/Python/3.9/lib/python/site-packages/iree/compiler/tools/../_mlir_libs/iree-lld" "--mlir-print-debuginfo" "--mlir-print-op-on-diagnostic=false" "--mlir-pass-pipeline-crash-reproducer=savedir/core-reproducer.mlir"
(lldb) run
Process 15264 launched: '___iree-install/bin/iree-compile' (x86_64)
Assertion failed: (use_empty() && "Cannot destroy a value that still has uses!"), function ~IRObjectWithUseList, file UseDefLists.h, line 175.
warning: iree-compile was compiled with optimization - stepping may behave oddly; variables may not be available.
Process 15264 stopped
* thread #17, stop reason = hit program assert
    frame #4: 0x0000000106139a83 iree-compile`mlir::Block::eraseArguments(llvm::function_ref<bool (mlir::BlockArgument)>) (.cold.1) at Block.cpp:175:5 [opt]
   165 	  // Update the cached position for all the arguments after the newly inserted
   166 	  // one.
   167 	  ++index;
   168 	  for (BlockArgument arg : llvm::drop_begin(arguments, index))
   169 	    arg.setArgNumber(index++);
   170 	  return arg;
   171 	}
   172 	
   173 	/// Insert one value to the given position of the argument list. The existing
   174 	/// arguments are shifted. The block is expected not to have predecessors.
-> 175 	BlockArgument Block::insertArgument(args_iterator it, Type type, Location loc) {
   176 	  assert(getPredecessors().empty() &&
   177 	         "cannot insert arguments to blocks with predecessors");
   178 	  return insertArgument(it->getArgNumber(), type, loc);
   179 	}
   180 	
   181 	void Block::eraseArgument(unsigned index) {
   182 	  assert(index < arguments.size());
   183 	  arguments[index].destroy();
   184 	  arguments.erase(arguments.begin() + index);
   185 	  for (BlockArgument arg : llvm::drop_begin(arguments, index))
Target 0: (iree-compile) stopped.
```

[savedir.zip](https://github.com/iree-org/iree/files/9951250/savedir.zip)

### What component(s) does this issue relate to?

Compiler

### Version information

iree commit 
1735445a30b9b60e4523b4ea113ecbfee7490058
(2022-11-03)

**Comment by stellaraccident** [comment]
This will be hard to diagnose without a reproducer. With any luck, one was generated. Can you poke around on your system and see if this flag dumped a file:` --mlir-pass-pipeline-crash-reproducer=savedir/core-reproducer.mlir`

We're still trying to make sure all of the ways that the compiler can get invoked produces the right debugging output on failure. If this didn't work, there is a magic environment variable we can give you which dumps exhaustive debugging information.

**Comment by benvanik** [comment]
This is fixed by #11099. There's a subsequent codegen error though that you'll want to file another issue for:
```
/Users/wiggles/Library/Python/3.9/lib/python/site-packages/veros/core/numerics.py:37:0: error: 'vector.gather' op operand #2 must be vector of integer or index values, but got 'index'
// -----// IR Dump After LinalgStrategyVectorizePass Failed (iree-linalg-strategy-vectorize-pass) //----- //
"func.func"() ({
  %0 = "arith.constant"() {value = dense<true> : vector<1x2xi1>} : () -> vector<1x2xi1>
  %1 = "arith.constant"() {value = dense<0.000000e+00> : vector<1x2xf32>} : () -> vector<1x2xf32>
  %2 = "arith.constant"() {value = 2 : index} : () -> index
  %3 = "arith.constant"() {value = 0 : index} : () -> index
  %4 = "arith.constant"() {value = 832 : index} : () -> index
  %5 = "arith.constant"() {value = 41 : index} : () -> index
  %6 = "hal.interface.binding.subspan"(%3) {alignment = 64 : index, binding = 0 : index, descriptor_type = 7 : i32, operand_segment_sizes = array<i32: 1, 0>, set = 0 : index} : (index) -> !flow.dispatch.tensor<readonly:tensor<44xf32>>
  %7 = "hal.interface.binding.subspan"(%4) {alignment = 64 : index, binding = 1 : index, descriptor_type = 7 : i32, operand_segment_sizes = array<i32: 1, 0>, set = 0 : index} : (index) -> !flow.dispatch.tensor<writeonly:tensor<1x2xf32>>
  %8 = "flow.dispatch.tensor.load"(%6) {operand_segment_sizes = array<i32: 1, 0, 0, 0, 0>, static_offsets = [0], static_sizes = [44], static_strides = [1]} : (!flow.dispatch.tensor<readonly:tensor<44xf32>>) -> tensor<44xf32>
  %9 = "hal.interface.workgroup.id"() {dimension = 0 : index} : () -> index
  %10 = "hal.interface.workgroup.count"() {dimension = 0 : index} : () -> index
  %11 = "affine.apply"(%9) {map = affine_map<()[s0] -> (s0 * 2)>} : (index) -> index
  %12 = "affine.apply"(%10) {map = affine_map<()[s0] -> (s0 * 2)>} : (index) -> index
  "scf.for"(%11, %2, %12) ({
  ^bb0(%arg0: index):
    %13 = "flow.dispatch.tensor.load"(%7, %arg0) {operand_segment_sizes = array<i32: 1, 0, 1, 0, 0>, static_offsets = [0, -9223372036854775808], static_sizes = [1, 2], static_strides = [1, 1]} : (!flow.dispatch.tensor<writeonly:tensor<1x2xf32>>, index) -> tensor<1x2xf32>
    %14 = "vector.gather"(%8, %3, %5, %0, %1) : (tensor<44xf32>, index, index, vector<1x2xi1>, vector<1x2xf32>) -> vector<1x2xf32>
    %15 = "vector.transfer_write"(%14, %13, %3, %3) {in_bounds = [true, true], operand_segment_sizes = array<i32: 1, 1, 2, 0>, permutation_map = affine_map<(d0, d1) -> (d0, d1)>} : (vector<1x2xf32>, tensor<1x2xf32>, index, index) -> tensor<1x2xf32>
    "flow.dispatch.tensor.store"(%15, %7, %arg0) {operand_segment_sizes = array<i32: 1, 1, 0, 1, 0, 0>, static_offsets = [0, -9223372036854775808], static_sizes = [1, 2], static_strides = [1, 1]} : (tensor<1x2xf32>, !flow.dispatch.tensor<writeonly:tensor<1x2xf32>>, index) -> ()
    "scf.yield"() : () -> ()
  }) : (index, index, index) -> ()
  "func.return"() : () -> ()
}) {function_type = () -> (), sym_name = "_main_dispatch_2_generic_1x2"} : () -> ()
```
